File: backend/app/services/calendar_service.py
"""
Calendar Service
Generates calendar events for transactions and expenses
"""

from sqlalchemy.orm import Session
from uuid import UUID
from datetime import datetime, date, timedelta
import calendar as calendar_module
import logging

from app.models import Transaction
from app.services.transaction_service import TransactionService

logger = logging.getLogger(__name__)


class CalendarService:
    """Calendar service for event generation."""

    @staticmethod
    def get_month_events(db: Session, user_id: UUID, year: int, month: int) -> dict:
        """Get all events for a month."""
        # Get all pending transactions for the month
        transactions = db.query(Transaction).filter(
            Transaction.user_id == user_id,
            Transaction.status == "PENDING",
        ).all()

        events = {}
        overdue = []

        today = date.today()

        for trans in transactions:
            if trans.expected_return_date:
                # Only include events in the requested month
                if trans.expected_return_date.year == year and trans.expected_return_date.month == month:
                    day = trans.expected_return_date.day

                    if day not in events:
                        events[day] = []

                    # Determine if overdue
                    is_overdue = trans.expected_return_date < today
                    overdue_days = (today - trans.expected_return_date).days if is_overdue else 0

                    events[day].append({
                        "id": str(trans.id),
                        "person": trans.person_name,
                        "amount": trans.amount,
                        "type": trans.transaction_type,
                        "purpose": trans.purpose,
                        "date": str(trans.expected_return_date),
                        "is_overdue": is_overdue,
                        "overdue_days": overdue_days,
                    })

                    # Track overdue
                    if is_overdue:
                        overdue.append({
                            "person": trans.person_name,
                            "amount": trans.amount,
                            "type": trans.transaction_type,
                            "overdue_days": overdue_days,
                        })

        return {
            "events": events,
            "overdue": overdue,
            "year": year,
            "month": month,
            "month_name": calendar_module.month_name[month],
        }

    @staticmethod
    def get_upcoming_events(db: Session, user_id: UUID, days_ahead: int = 30) -> dict:
        """Get upcoming events for the next N days."""
        today = date.today()
        future_date = today + timedelta(days=days_ahead)

        transactions = db.query(Transaction).filter(
            Transaction.user_id == user_id,
            Transaction.status == "PENDING",
            Transaction.expected_return_date >= today,
            Transaction.expected_return_date <= future_date,
        ).order_by(Transaction.expected_return_date).all()

        events = []
        for trans in transactions:
            days_until = (trans.expected_return_date - today).days
            events.append({
                "id": str(trans.id),
                "person": trans.person_name,
                "amount": trans.amount,
                "type": trans.transaction_type,
                "date": str(trans.expected_return_date),
                "days_until": days_until,
                "is_today": days_until == 0,
                "is_tomorrow": days_until == 1,
            })

        return {
            "events": events,
            "days_ahead": days_ahead,
            "total_events": len(events),
        }

    @staticmethod
    def get_overdue_events(db: Session, user_id: UUID) -> list[dict]:
        """Get all overdue pending transactions."""
        today = date.today()

        transactions = db.query(Transaction).filter(
            Transaction.user_id == user_id,
            Transaction.status == "PENDING",
            Transaction.expected_return_date < today,
        ).order_by(Transaction.expected_return_date).all()

        overdue_events = []
        for trans in transactions:
            overdue_days = (today - trans.expected_return_date).days
            overdue_events.append({
                "id": str(trans.id),
                "person": trans.person_name,
                "amount": trans.amount,
                "type": trans.transaction_type,
                "date": str(trans.expected_return_date),
                "overdue_days": overdue_days,
            })

        return overdue_events

    @staticmethod
    def get_calendar_summary(db: Session, user_id: UUID) -> dict:
        """Get summary of calendar events."""
        today = date.today()
        week_end = today + timedelta(days=7)
        month_end = today.replace(day=1) + timedelta(days=32)
        month_end = month_end.replace(day=1) - timedelta(days=1)

        # Upcoming this week
        week_events = db.query(Transaction).filter(
            Transaction.user_id == user_id,
            Transaction.status == "PENDING",
            Transaction.expected_return_date >= today,
            Transaction.expected_return_date <= week_end,
        ).all()

        # Upcoming this month
        month_events = db.query(Transaction).filter(
            Transaction.user_id == user_id,
            Transaction.status == "PENDING",
            Transaction.expected_return_date >= today,
            Transaction.expected_return_date <= month_end,
        ).all()

        # Overdue
        overdue = db.query(Transaction).filter(
            Transaction.user_id == user_id,
            Transaction.status == "PENDING",
            Transaction.expected_return_date < today,
        ).all()

        return {
            "this_week": {
                "count": len(week_events),
                "total": sum(t.amount for t in week_events),
            },
            "this_month": {
                "count": len(month_events),
                "total": sum(t.amount for t in month_events),
            },
            "overdue": {
                "count": len(overdue),
                "total": sum(t.amount for t in overdue),
            },
        }

File: frontend/src/services/calendarService.ts
import { supabase } from './supabaseClient'
import { format, differenceInDays } from 'date-fns'

export interface CalendarEvent {
  id:          string
  person:      string
  amount:      number
  type:        'BORROWED' | 'LENT'
  purpose:     string
  date:        string
  is_overdue:  boolean
  overdue_days: number
}

export interface OverdueEvent {
  id:          string
  person:      string
  amount:      number
  type:        'BORROWED' | 'LENT'
  date:        string
  overdue_days: number
}

async function getUserId(): Promise<string> {
  const { data: { user }, error } = await supabase.auth.getUser()
  if (error || !user) throw new Error('Not authenticated')
  return user.id
}

const MONTH_NAMES = [
  'January','February','March','April','May','June',
  'July','August','September','October','November','December',
]

class CalendarService {
  async getMonthEvents(year: number, month: number) {
    const userId = await getUserId()
    const from   = `${year}-${String(month).padStart(2, '0')}-01`
    const daysInMonth = new Date(year, month, 0).getDate()
    const to     = `${year}-${String(month).padStart(2, '0')}-${String(daysInMonth).padStart(2, '0')}`
    const today  = new Date().toISOString().split('T')[0]

    const { data, error } = await supabase
      .from('transactions')
      .select('id, transaction_type, person_name, amount, purpose, expected_return_date, status')
      .eq('user_id', userId)
      .eq('status', 'PENDING')
      .not('expected_return_date', 'is', null)
      .gte('expected_return_date', from)
      .lte('expected_return_date', to)

    if (error) throw error

    const events: Record<number, CalendarEvent[]> = {}
    const overdue: Array<{ person: string; amount: number; type: 'BORROWED' | 'LENT'; overdue_days: number }> = []

    for (const row of data || []) {
      const day = new Date(row.expected_return_date).getDate()
      const isOverdue = row.expected_return_date < today
      const overdueDays = isOverdue
        ? differenceInDays(new Date(today), new Date(row.expected_return_date))
        : 0

      if (!events[day]) events[day] = []
      events[day].push({
        id:          row.id,
        person:      row.person_name,
        amount:      Number(row.amount),
        type:        row.transaction_type,
        purpose:     row.purpose || '',
        date:        row.expected_return_date,
        is_overdue:  isOverdue,
        overdue_days: overdueDays,
      })

      if (isOverdue) {
        overdue.push({ person: row.person_name, amount: Number(row.amount), type: row.transaction_type, overdue_days: overdueDays })
      }
    }

    return {
      events,
      overdue,
      year,
      month,
      month_name: MONTH_NAMES[month - 1],
    }
  }

  async getOverdueEvents(): Promise<OverdueEvent[]> {
    const userId = await getUserId()
    const today  = new Date().toISOString().split('T')[0]

    const { data, error } = await supabase
      .from('transactions')
      .select('id, transaction_type, person_name, amount, expected_return_date')
      .eq('user_id', userId)
      .eq('status', 'PENDING')
      .not('expected_return_date', 'is', null)
      .lt('expected_return_date', today)
      .order('expected_return_date', { ascending: true })

    if (error) throw error

    return (data || []).map((row) => ({
      id:          row.id,
      person:      row.person_name,
      amount:      Number(row.amount),
      type:        row.transaction_type,
      date:        row.expected_return_date,
      overdue_days: differenceInDays(new Date(today), new Date(row.expected_return_date)),
    }))
  }

  async getUpcomingEvents(daysAhead: number = 30) {
    const userId = await getUserId()
    const today  = new Date().toISOString().split('T')[0]
    const future = new Date(); future.setDate(future.getDate() + daysAhead)
    const to     = future.toISOString().split('T')[0]

    const { data, error } = await supabase
      .from('transactions')
      .select('id, transaction_type, person_name, amount, expected_return_date')
      .eq('user_id', userId)
      .eq('status', 'PENDING')
      .not('expected_return_date', 'is', null)
      .gte('expected_return_date', today)
      .lte('expected_return_date', to)
      .order('expected_return_date', { ascending: true })

    if (error) throw error

    const events = (data || []).map((row) => {
      const daysUntil = differenceInDays(new Date(row.expected_return_date), new Date(today))
      return {
        id:          row.id,
        person:      row.person_name,
        amount:      Number(row.amount),
        type:        row.transaction_type,
        date:        row.expected_return_date,
        days_until:  daysUntil,
        is_today:    daysUntil === 0,
        is_tomorrow: daysUntil === 1,
      }
    })

    return { events, days_ahead: daysAhead, total_events: events.length }
  }

  async getCalendarSummary() {
    const userId = await getUserId()
    const today  = new Date()
    const todayStr = today.toISOString().split('T')[0]
    const weekEnd  = new Date(today); weekEnd.setDate(weekEnd.getDate() + 7)
    const monthEnd = new Date(today.getFullYear(), today.getMonth() + 1, 0)

    const { data } = await supabase
      .from('transactions')
      .select('amount, expected_return_date, status')
      .eq('user_id', userId)
      .eq('status', 'PENDING')
      .not('expected_return_date', 'is', null)

    const rows = data || []
    const thisWeek  = rows.filter((r) => r.expected_return_date >= todayStr && r.expected_return_date <= weekEnd.toISOString().split('T')[0])
    const thisMonth = rows.filter((r) => r.expected_return_date >= todayStr && r.expected_return_date <= monthEnd.toISOString().split('T')[0])
    const overdue   = rows.filter((r) => r.expected_return_date < todayStr)

    const sum = (arr: any[]) => arr.reduce((s, r) => s + Number(r.amount), 0)

    return {
      this_week:  { count: thisWeek.length,  total: sum(thisWeek) },
      this_month: { count: thisMonth.length, total: sum(thisMonth) },
      overdue:    { count: overdue.length,   total: sum(overdue) },
    }
  }
}

export default new CalendarService()
